### PR TITLE
Optimize environment setup script to speed up test time.

### PR DESCRIPTION
**Pull request by henrywang**
1. Enable docker cache to speed up container build time.

### CODE REVIEW

**Comment by henrywang** [comment]
[Optimize welder-web e2e build and execution times](https://trello.com/c/g4qQmiIa/397-optimize-welder-web-e2e-build-and-execution-times)

File: .travis.yml [hunk 1]
```diff
@@ -5,17 +5,27 @@ services:
 language: node_js
 node_js:
   - "6"
+env:
+  global:
+    - CACHE_DIR=$HOME/.cache/docker
```
Comment (atodorov) [inline .travis.yml]: Shouldn't this be named DOCKER_DIR or similar ?

File: .travis.yml [hunk 2]
```diff
@@ -5,17 +5,27 @@ services:
...
-  - npm install
-  - node run build
-env:
-  - COMMAND=eslint
-  - COMMAND=stylelint
-  - COMMAND=unit-test
-  - COMMAND=end-to-end-test
+  - if [ "$COMMAND" == "end-to-end-test" ]; then make e2e-before-install; fi
+install:
+  - if [ "$COMMAND" == "end-to-end-test" ]; then make e2e-install; fi
```
Comment (atodorov) [inline .travis.yml]: You can instead make the end-to-end make target depend on these other targets:

`end-to-end-test: e2e-before-install e2e-install`

File: Makefile [hunk 3]
```diff
@@ -4,26 +4,39 @@ eslint:
...
+e2e-before-install:
+	if [ -f ${CACHE_FILE_FEDORA} ]; then gunzip -c ${CACHE_FILE_FEDORA} | sudo docker load; fi
+	if [ -f ${CACHE_FILE_WEB} ]; then gunzip -c ${CACHE_FILE_WEB} | sudo docker load; fi
+	if [ -f ${CACHE_FILE_API} ]; then gunzip -c ${CACHE_FILE_API} | sudo docker load; fi
+	if [ -f ${CACHE_FILE_END_TO_END} ]; then gunzip -c ${CACHE_FILE_END_TO_END} | sudo docker load; fi
+	git clone https://github.com/weldr/welder-deployment.git
+	git clone https://github.com/weldr/bdcs-api-rs.git
+	wget https://s3.amazonaws.com/weldr/metadata.db
+	sudo docker network create welder
+	mv welder-deployment/Dockerfile-weld-f25 welder-deployment/Dockerfile
```
Comment (atodorov) [inline Makefile]: I feel like the above 5 lines should go into an else block somewhere. Probably for the if on line 23.

Tip if you want to write a multi line script in Makefile then use \ and ; characters.

File: Makefile [hunk 4]
```diff
@@ -4,26 +4,39 @@ eslint:
...
+	sudo docker build --cache-from weld/end-to-end:latest -t weld/end-to-end:latest test/end-to-end/
+	sudo docker build --cache-from weld/web:latest -t weld/web:latest .
+	sudo docker build --cache-from weld/bdcs-api:latest -t weld/bdcs-api:latest bdcs-api-rs/
+	sudo docker run -d --name api --restart=always -p 4000:4000 -v bdcs-recipes-volume:/bdcs-recipes -v `pwd`:/mddb --network welder --security-opt label=disable weld/bdcs-api
+	sudo docker run -d --name web --restart=always -p 80:3000 --network welder weld/web
+	mkdir -p $CACHE_DIR
+	if [[ ${TRAVIS_BRANCH} == "master" ]] && [[ ${TRAVIS_PULL_REQUEST} == "false" ]]; then sudo docker save weld/fedora:25 | gzip > ${CACHE_FILE_FEDORA}; fi
+	if [[ ${TRAVIS_BRANCH} == "master" ]] && [[ ${TRAVIS_PULL_REQUEST} == "false" ]]; then sudo docker save weld/web:latest | gzip > ${CACHE_FILE_WEB}; fi
+	if [[ ${TRAVIS_BRANCH} == "master" ]] && [[ ${TRAVIS_PULL_REQUEST} == "false" ]]; then sudo docker save weld/bdcs-api:latest | gzip > ${CACHE_FILE_API}; fi
+	if [[ ${TRAVIS_BRANCH} == "master" ]] && [[ ${TRAVIS_PULL_REQUEST} == "false" ]]; then sudo docker save weld/end-to-end:latest | gzip > ${CACHE_FILE_END_TO_END}; fi
```
Comment (atodorov) [inline Makefile]: These need to go into .travis.yml, after_success section. No CI dependent commands into Makefile.

File: Makefile [hunk 5]
```diff
@@ -4,26 +4,39 @@ eslint:
...
+	git clone https://github.com/weldr/bdcs-api-rs.git
+	wget https://s3.amazonaws.com/weldr/metadata.db
+	sudo docker network create welder
+	mv welder-deployment/Dockerfile-weld-f25 welder-deployment/Dockerfile
+
+e2e-install:
+	sudo docker build --cache-from weld/fedora:25 -t weld/fedora:25 welder-deployment/
+	sudo docker build --cache-from weld/end-to-end:latest -t weld/end-to-end:latest test/end-to-end/
+	sudo docker build --cache-from weld/web:latest -t weld/web:latest .
+	sudo docker build --cache-from weld/bdcs-api:latest -t weld/bdcs-api:latest bdcs-api-rs/
```
Comment (atodorov) [inline Makefile]: Note: we need to test a build job when some of the backend code has changed before we merge this change.

File: Makefile (hunk 4 above)
Comment (atodorov) [inline Makefile]: Note: on second discussion these `if TRAVIS_BRANCH == master` statements look to not be necessary b/c they don't really help except prevent a few docker-save commands from running. Until we have a second build job (with absolutely the same code) to compare gainst there is nothing we can discuss about.

File: Makefile [hunk 6]
```diff
@@ -4,26 +4,39 @@ eslint:
 stylelint:
 	npm run stylelint
 
+build-test:
+	npm install
+	npm run build
```
Comment (atodorov) [inline Makefile]: The deleted version from .travis.yml is `node run build` instead of `npm run build`. What is the difference between the two?

Comment (henrywang) [inline Makefile]: Just follow the same npm style, beyond that there's no difference here.

**Comment by atodorov** [comment]
Closing in favor of #132

**Pull request closed without merging**